### DISCUSSION TITLE
Doubt about the error message "An unexpected error occurred"

### DISCUSSION THREAD

**Discussion by aerdnach** [directus/directus#13083] [Archive]
I have a Collection A with two fields: id and a text field named fieldA. Also I have another collection with three fields: an id, a text field named fieldB, and a relational field named retationWithA as a M2O relation with collection A; moreover on this field i set a PREVENT relational trigger, so that a deletion of an item in collection A is prevented if there is at least one item in collection B that references it.

With this in mind, I would like to explain my doubt. If I delete an item in Collection A, which is referenced by an item in Collection B, with a user with Administrator role I get an error 500 with the following message 

"[INTERNAL_SERVER_ERROR] delete from "Collection_A" where "id" in ($1) - update or delete on table "Collection_A" violates foreign key constraint "collection_b_retationwitha_foreign" on table "Collection_B". 

This way I know that if I want to delete the item in collection A I must first delete the items that reference it in collection B. On the other hand, if I make a deletion with a role that does not have administrator rights an error 500 with the message "[INTERNAL_SERVER_ERROR] An unexpected error occurred", without any details on what happened, and without being able to remedy what happened, considering also that it could be the non-administrator user who created the collections, and should know that he is violating a constraint that he has defined.

What I would expect is for the non-admin user to receive the same error as the admin user, so that he or she can perform the necessary actions to fix the problem. I also think that a 5xx error (server error) would not be appropriate in such a scenario, but that a 4xx error (client error) would be more appropriate (perhaps a [409 error](https://datatracker.ietf.org/doc/html/rfc7231#section-6.5.8)), since the error is caused by a user operation that violates a constraint in the data model that the user has defined.

This was a long post. Thanks for your attention :)

**Comment by azrikahar** [comment]
There a few discussions/issues revolving around this topic, but we are trying to converge them over at #4368 👍

> What I would expect is for the non-admin user to receive the same error as the admin user, so that he or she can perform the necessary actions to fix the problem.

Agreed that this is a better experience! It's just that the current implementation started out with the notion of preventing potential over-exposing schema/data that may be sensitive, as the aforementioned discussion described. We can/should definitely fit in what you proposed (especially more appropriate errors) when that discussion gets implemented or fleshed out further.

**Comment by aerdnach** [comment]
Here is my proposal for a quick enhancement. We could introduce a boolean environment variable, which we could call SHOW_ERROR_DETAILS, that if set to TRUE makes it possible to show error details even to non-administrator users. This variable would be used in the api/src/middleware/error-handler.ts file on line 64, setting it to OR with the control already in place. I tested it locally and it works correctly.

If you find this type of enhancement interesting, I could submit a pull request.

**Reply by rijkvanzanten** [reply]
Sharing the unexpected error messages with the outside world can be considered a security problem, so I'm not sure if that's something we'd want to expose as an option. For your particular error example above, we should abstract away the unexpected database error into a proper exception type and return it in a known format 👍🏻
